app/grandchallenge/components/tasks.py

```diff
@@ -1249,17 +1209,35 @@ def add_image_to_object(
         try:
             civ.full_clean()
         except ValidationError as e:
-            # this should only happen for new uploads
-            upload_session.status = RawImageUploadSession.FAILURE
-            upload_session.error_message = format_validation_error_message(
-                error=e
+            upload_session.update_status(
+                status=RawImageUploadSession.FAILURE,
+                error_message=f"File for interface {interface.title} failed validation: {format_validation_error_message(e)}",
+                linked_object=object,
+            )
+            return
+        except Exception as e:
+            upload_session.update_status(
+                status=RawImageUploadSession.FAILURE,
+                error_message="An unexpected error occurred",
+                linked_object=object,
             )
-            upload_session.save()
+            logger.error(e, exc_info=True)
             return
-    object.values.add(civ)
 
+    try:
+        object.remove_civ(civ=current_value)
+        object.add_civ(civ=civ)
+    except RuntimeError as e:
+        # for Jobs this happens when validation for another input failed
+        # and the job status was updated to CANCELLED
+        logger.error(e, exc_info=True)
+        return
 
-@acks_late_micro_short_task
+    if linked_task is not None:
+        on_commit(signature(linked_task).apply_async)
+
+
+@acks_late_micro_short_task(retry_on=(LockNotAcquiredException,))
```

Comment:
I'm still a little bit concerned about competing locks for jobs with multiple file inputs for both `add_file_to_object` and `add_image_to_object`. Should we set `delayed_retry=False`?

Reply:
I had to read up on that argument first, but yes, I think you're right that setting `delayed_retry=False` is the right thing to do here. 

If I understood things correctly then this means that a task will be retried on the same queue (not the delayed queue) and will be immediately re-queued rather than with a delay of 5 minutes (as it would on the delayed queue). I was a bit worried that retrying immediately would be pointless if there is a blocking task, but by the time the re-queued task gets picked up, the blocking tasks will most likely have finished already anyway, and if not, it can be retried again, we allow plenty of retries.

Reply:
Clear, thanks for explaining.